test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/io/CharArrayWriterTest.scala

```diff
@@ -0,0 +1,227 @@
+    */
+  @throws[IOException]
+  @Test def test_reset(): Unit = {
+    cw.write("HelloWorld", 5, 5)
+    cw.reset()
+    cw.write("HelloWorld", 0, 5)
+    cr = new CharArrayReader(cw.toCharArray)
+    val c = new Array[Char](100)
+    cr.read(c, 0, 5)
+    assertEquals("Reset failed to reset buffer", "Hello", new String(c, 0, 5))
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#size()
+    */
+  @Test def test_size(): Unit = {
+    assertEquals("Returned incorrect size", 0, cw.size)
+    cw.write(hw, 5, 5)
+    assertEquals("Returned incorrect size", 5, cw.size)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#toCharArray()
+    */
+  @throws[IOException]
+  @Test def test_toCharArray(): Unit = {
+    cw.write("HelloWorld", 0, 10)
+    cr = new CharArrayReader(cw.toCharArray)
+    val c = new Array[Char](100)
+    cr.read(c, 0, 10)
+    assertEquals("toCharArray failed to return correct array", "HelloWorld", new String(c, 0, 10))
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#toString()
+    */
+  @Test def test_toString(): Unit = {
+    cw.write("HelloWorld", 5, 5)
+    cr = new CharArrayReader(cw.toCharArray)
+    assertEquals("Returned incorrect string", "World", cw.toString)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#write(char[], int, int)
+    */
+  @throws[IOException]
+  @Test def test_write$CII(): Unit = {
+    cw.write(hw, 5, 5)
+    cr = new CharArrayReader(cw.toCharArray)
+    val c = new Array[Char](100)
+    cr.read(c, 0, 5)
+    assertEquals("Writer failed to write correct chars", "World", new String(c, 0, 5))
+  }
+
+  @Test def test_write$CII_2(): Unit = { // Regression for HARMONY-387
+    val obj = new CharArrayWriter
+    try {
+      obj.write(Array[Char]('0'), 0, -1)
+      fail("IndexOutOfBoundsException expected")
+    } catch {
+      case t: IndexOutOfBoundsException =>
+        assertEquals("IndexOutOfBoundsException rather than a subclass expected", classOf[IndexOutOfBoundsException], t.getClass)
+    }
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#write(int)
+    */
+  @throws[IOException]
+  @Test def test_writeI(): Unit = {
+    cw.write('T')
+    cr = new CharArrayReader(cw.toCharArray)
+    assertEquals("Writer failed to write char", 'T', cr.read)
+  }
+
+  /**
+    * @tests java.io.CharArrayWriter#write(java.lang.String, int, int)
+    */
+  @throws[IOException]
+  @Test def test_writeLjava_lang_StringII(): Unit = {
+    cw.write("HelloWorld", 5, 5)
+    cr = new CharArrayReader(cw.toCharArray)
+    val c = new Array[Char](100)
+    cr.read(c, 0, 5)
+    assertEquals("Writer failed to write correct chars", "World", new String(c, 0, 5))
+  }
+
+  @throws[StringIndexOutOfBoundsException]
+  @Test def test_writeLjava_lang_StringII_2(): Unit = {
+    val obj = new CharArrayWriter
+    try {
+      obj.write(null.asInstanceOf[String], -1, 0)
+      fail("NullPointerException expected")
+    } catch {
+      case _: NullPointerException => // Expected
```

Comment:
We typically don't write tests for expected NPEs because NPEs are Undefined Behavior in Scala.js.